File: server/internal/serverstore/usage_dept.go
package serverstore

import (
	"database/sql"
	"sort"
	"strconv"
	"strings"
)

// ---------------------------------------------------------------------------
// 部门维度用量聚合(2026-09 用量中心重构, 设计 docs/decisions/2026-09-02-usage-center-redesign.md)
//
// 口径与部门预算 enforcement 完全一致:部门树内合计 = 成员归属祖先链全部计入,
// 同一用户对同一部门只计一次(去重)。「未归属任何部门的用量」不计入部门视图
// (总览/成员维度仍全额可见)。
// ---------------------------------------------------------------------------

// RegroupByDept 把 group=user 的聚合行按部门树归并为按部门行(树先序排序)。
// rows.Label 为用户名(或已删用户的 user_id 字符串)。
func RegroupByDept(db *sql.DB, rows []UsageAggregateRow) ([]UsageAggregateRow, error) {
	nodes, err := loadGroupTree(db)
	if err != nil {
		return nil, err
	}
	if len(nodes) == 0 {
		return nil, nil
	}
	// 用户(可能多部门) → 所属部门名集合(已祖先展开、去重)
	userDepts, err := userIDToDepts(db, nodes)
	if err != nil {
		return nil, err
	}
	// label → userID:用户名优先,否则按 user_id 字符串解析
	labelToUID := map[string]int64{}
	{ // 填充原用户名(仅对已归属部门用户必要)
		rowsRes, err := db.Query(`SELECT id, username FROM users`)
		if err != nil {
			return nil, err
		}
		defer rowsRes.Close()
		for rowsRes.Next() {
			var id int64
			var name string
			if err := rowsRes.Scan(&id, &name); err != nil {
				return nil, err
			}
			labelToUID[name] = id
		}
		if err := rowsRes.Err(); err != nil {
			return nil, err
		}
	}

	agg := map[string]*UsageAggregateRow{}
	for _, r := range rows {
		uid, ok := labelToUID[r.Label]
		if !ok {
			if n, err := strconv.ParseInt(r.Label, 10, 64); err == nil {
				uid, ok = n, true
			}
		}
		if !ok {
			continue // 未归属部门(或已删且无法解析)
		}
		for _, dept := range userDepts[uid] {
			cur := agg[dept]
			if cur == nil {
				cp := r
				cp.Label = dept // 首行作为种子,label 换成部门名
				cur = &cp
				agg[dept] = cur
			} else {
				cur.PromptTokens += r.PromptTokens
				cur.CompletionTokens += r.CompletionTokens
				cur.Requests += r.Requests
				cur.EmbedRequests += r.EmbedRequests
				cur.EmbedTokens += r.EmbedTokens
				cur.CacheTokens += r.CacheTokens
				cur.Cost += r.Cost
			}
		}
	}
	out := make([]UsageAggregateRow, 0, len(agg))
	for _, r := range agg {
		out = append(out, *r)
	}
	// 树先序排序:父部门在前,子部门跟随(展示层级感)
	order := map[string]int{}
	for i, n := range preOrderNodes(nodes) {
		order[n.name] = i
	}
	sort.Slice(out, func(i, j int) bool {
		oi, oiOK := order[out[i].Label]
		oj, ojOK := order[out[j].Label]
		if !oiOK || !ojOK {
			return out[i].Label < out[j].Label
		}
		return oi < oj
	})
	return out, nil
}

// DeptUserIDsByName 部门名(含其子树)的成员 user_id 集合——与 DeptMemberIDs 同语义,
// 供 SQL 过滤(WithDept)使用。
func DeptUserIDsByName(db *sql.DB, dept string) ([]int64, error) {
	sub, err := deptSubtreeIDs(db, dept)
	if err != nil {
		return nil, err
	}
	placeholders := strings.Repeat("?,", len(sub))
	placeholders = placeholders[:len(placeholders)-1]
	args := make([]any, 0, len(sub))
	for _, id := range sub {
		args = append(args, id)
	}
	rows, err := db.Query(`SELECT DISTINCT user_id FROM user_groups WHERE group_id IN (`+placeholders+`)`, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []int64
	for rows.Next() {
		var uid int64
		if err := rows.Scan(&uid); err != nil {
			return nil, err
		}
		out = append(out, uid)
	}
	return out, rows.Err()
}

// deptSubtreeIDs 返回部门名对应的子树 group id 列表(含自身)。
func deptSubtreeIDs(db *sql.DB, dept string) ([]int64, error) {
	nodes, err := loadGroupTree(db)
	if err != nil {
		return nil, err
	}
	byID := map[int64]groupNode{}
	for _, n := range nodes {
		byID[n.id] = n
	}
	n, ok := findNodeByName(byID, dept)
	if !ok {
		return nil, ErrNotFound
	}
	return subtreeGroupIDs(db, n.id)
}

// userIDToDepts 用户 → 部门名集合(成员归属祖先链展开, 去重)。
func userIDToDepts(db *sql.DB, nodes []groupNode) (map[int64][]string, error) {
	byID := map[int64]groupNode{}
	children := map[int64][]int64{}
	for _, n := range nodes {
		byID[n.id] = n
		children[n.parent] = append(children[n.parent], n.id)
	}
	ancestors := func(id int64) []string {
		out := []string{}
		cur := byID[id]
		for cur.id != 0 {
			out = append(out, cur.name)
			cur = byID[cur.parent]
			if cur.id == 0 {
				break
			}
		}
		return out
	}
	rows, err := db.Query(`SELECT user_id, group_id FROM user_groups`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := map[int64][]string{}
	for rows.Next() {
		var uid, gid int64
		if err := rows.Scan(&uid, &gid); err != nil {
			return nil, err
		}
		if _, ok := byID[gid]; !ok {
			continue
		}
		seen := map[string]bool{}
		for _, name := range ancestors(gid) {
			if !seen[name] {
				seen[name] = true
				out[uid] = append(out[uid], name)
			}
		}
	}
	return out, rows.Err()
}

// preOrderNodes 树先序(根在前, 其子按 id)。
func preOrderNodes(nodes []groupNode) []groupNode {
	children := map[int64][]int64{}
	rootID := int64(0)
	for _, n := range nodes {
		children[n.parent] = append(children[n.parent], n.id)
	}
	// 根 = 无父(或父不存在)的节点
	byID := map[int64]groupNode{}
	for _, n := range nodes {
		byID[n.id] = n
		if _, ok := byID[n.parent]; !ok {
			rootID = n.id
		}
	}
	out := []groupNode{}
	stack := []int64{rootID}
	for len(stack) > 0 {
		curID := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if n, ok := byID[curID]; ok {
			out = append(out, n)
			// 逆序压栈保证子部门按 id 正序弹出
			kids := children[curID]
			for i := len(kids) - 1; i >= 0; i-- {
				stack = append(stack, kids[i])
			}
		}
	}
	return out
}

File: rust/crates/dsh-store/src/usage_dept.rs
//! 部门维度用量聚合基础（Go `serverstore/usage_dept.go` 的核心函数等价）。

use crate::errors::{map_db_error, StoreError};
use sqlx::Row;

/// DeptUserIDsByName 部门名（含子树）的成员 user_id 集合——与 DeptMemberIDs 同语义。
pub async fn dept_user_ids_by_name(pool: &sqlx::PgPool, dept: &str) -> Result<Vec<i64>, StoreError> {
    let sub = dept_subtree_ids(pool, dept).await?;
    if sub.is_empty() {
        return Ok(Vec::new());
    }
    let mut sql = String::from("SELECT DISTINCT user_id FROM user_groups WHERE group_id IN (");
    for (i, _) in sub.iter().enumerate() {
        if i > 0 {
            sql.push(',');
        }
        sql.push_str(&format!("${}", i + 1));
    }
    sql.push(')');
    let mut q = sqlx::query(&sql);
    for id in &sub {
        q = q.bind(*id as i32);
    }
    let rows = q.fetch_all(pool).await.map_err(map_db_error)?;
    let mut out = Vec::new();
    for r in rows {
        out.push(r.get::<i32, _>("user_id") as i64);
    }
    Ok(out)
}

/// dept_subtree_ids 返回部门名对应子树 group id 列表（含自身）；部门不存在 → ErrNotFound。
pub async fn dept_subtree_ids(pool: &sqlx::PgPool, dept: &str) -> Result<Vec<i64>, StoreError> {
    let id: Option<i64> = sqlx::query_scalar("SELECT id FROM groups WHERE name = $1")
        .bind(dept)
        .fetch_optional(pool)
        .await
        .map_err(map_db_error)?;
    let Some(id) = id else {
        return Err(StoreError::NotFound);
    };
    crate::departments::subtree_group_ids(pool, id).await
}

/// user_id_to_depts 用户 → 部门名集合（成员归属祖先链展开，去重）。
/// 返回 map[user_id] -> Vec<dept name>（按祖先顺序）。
pub async fn user_id_to_depts(
    pool: &sqlx::PgPool,
) -> Result<std::collections::HashMap<i64, Vec<String>>, StoreError> {
    // 组织树
    let nodes: Vec<(i64, String, i32)> = sqlx::query_as("SELECT id, name, parent_id FROM groups")
        .fetch_all(pool)
        .await
        .map_err(map_db_error)?;
    let mut by_id: std::collections::HashMap<i64, (String, i64)> = std::collections::HashMap::new();
    for (id, name, parent) in &nodes {
        by_id.insert(*id, (name.clone(), *parent as i64));
    }
    // 用户组关系
    let rows = sqlx::query("SELECT user_id, group_id FROM user_groups")
        .fetch_all(pool)
        .await
        .map_err(map_db_error)?;
    let mut out: std::collections::HashMap<i64, Vec<String>> = std::collections::HashMap::new();
    for r in rows {
        let uid = r.get::<i32, _>("user_id") as i64;
        let gid = r.get::<i32, _>("group_id") as i64;
        if !by_id.contains_key(&gid) {
            continue;
        }
        let mut seen = std::collections::HashSet::new();
        let mut cur = gid;
        loop {
            let (name, parent) = by_id.get(&cur).cloned().unwrap_or_default();
            if name.is_empty() {
                break;
            }
            if seen.insert(name.clone()) {
                out.entry(uid).or_default().push(name);
            }
            if parent == 0 {
                break;
            }
            cur = parent;
        }
    }
    Ok(out)
}

/// pre_order_nodes 树先序（根在前，子按 id 升序）。
#[derive(Debug, Clone)]
pub struct DeptNode {
    pub id: i64,
    pub name: String,
    pub parent: i64,
}

/// pre_order_nodes_from_pool 装载树并先序排序。
pub async fn pre_order_nodes(pool: &sqlx::PgPool) -> Result<Vec<DeptNode>, StoreError> {
    let nodes: Vec<(i64, String, i32)> = sqlx::query_as("SELECT id, name, parent_id FROM groups")
        .fetch_all(pool)
        .await
        .map_err(map_db_error)?;
    let mut by_id: std::collections::HashMap<i64, DeptNode> = std::collections::HashMap::new();
    let mut children: std::collections::HashMap<i64, Vec<i64>> = std::collections::HashMap::new();
    for (id, name, parent) in &nodes {
        by_id.insert(*id, DeptNode { id: *id, name: name.clone(), parent: *parent as i64 });
        children.entry(*parent as i64).or_default().push(*id);
        if !nodes.iter().any(|(_, _, p)| *p as i64 == *id) {
            // 无父引用 = 根（近似：处理孤立）
        }
    }
    // 根 = 无父的节点
    let mut roots = Vec::new();
    for (id, node) in &by_id {
        if node.parent == 0 || !by_id.contains_key(&node.parent) {
            roots.push(*id);
        }
    }
    roots.sort();
    let mut out = Vec::new();
    let mut stack = roots;
    while let Some(cur) = stack.pop() {
        if let Some(n) = by_id.get(&cur) {
            out.push(n.clone());
            if let Some(kids) = children.get(&cur) {
                let mut kids = kids.clone();
                kids.sort();
                for k in kids.into_iter().rev() {
                    stack.push(k);
                }
            }
        }
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testutil::new_test_db;

    #[tokio::test]
    async fn dept_users_by_name() {
        let pool = new_test_db().await;
        let uid = crate::users::create_user(
            &pool,
            &crate::users::User {
                username: "alice".into(),
                source: "local".into(),
                status: 1,
                ..Default::default()
            },
        )
        .await
        .unwrap();
        let dept = crate::departments::create_department(&pool, "研发部", 0, 0, "").await.unwrap();
        crate::groups::add_user_group(&pool, uid, dept).await.unwrap();
        let ids = dept_user_ids_by_name(&pool, "研发部").await.unwrap();
        assert!(ids.contains(&uid));
        // 不存在 → ErrNotFound
        assert_eq!(
            dept_user_ids_by_name(&pool, "不存在").await.unwrap_err(),
            StoreError::NotFound
        );
    }

    #[tokio::test]
    async fn user_to_depts() {
        let pool = new_test_db().await;
        let uid = crate::users::create_user(
            &pool,
            &crate::users::User {
                username: "bob".into(),
                source: "local".into(),
                status: 1,
                ..Default::default()
            },
        )
        .await
        .unwrap();
        let parent = crate::departments::create_department(&pool, "公司", 0, 0, "").await.unwrap();
        let child = crate::departments::create_department(&pool, "研发", parent, 0, "").await.unwrap();
        crate::groups::add_user_group(&pool, uid, child).await.unwrap();
        let m = user_id_to_depts(&pool).await.unwrap();
        let depts = m.get(&uid).unwrap();
        assert!(depts.contains(&"公司".to_string()) && depts.contains(&"研发".to_string()));
    }
}
